elsa-types.el

```diff
@@ -476,5 +476,125 @@ other, then this is a supertype of other."
 (cl-defmethod elsa-type-describe ((this elsa-generic-type))
   (symbol-name (oref this label)))
 
+;; One-dimensional sparse arrays indexed by characters
+(defclass elsa-type-charttable (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-charttable))
+  "Charttable")
+
+;; One-dimensional arrays of t or nil.
+(defclass elsa-type-boolvector (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-boolvector))
+  "Boolvector")
+
+;; Super-fast lookup tables
+(defclass elsa-type-hashtable (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-hashtable))
+  "Hashtable")
+
+;; A piece of executable code you can call from elsewhere.
+(defclass elsa-type-function (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-function))
+  "Function")
+
+;; A method of expanding an expression into another expression, more fundamental but less pretty
+(defclass elsa-type-macro (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-macro))
+  "Macro")
+
+;; A function written in C, callable from Lisp
+(defclass elsa-type-primitive (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-primitive))
+  "Primitive")
+
+;; function written in Lisp, then compiled
+(defclass elsa-type-bytecode (elsa-type) ())
```

Comment:
Same as `macro`.  If there is no type-level distinction we can merge this with the function type, if there is then we should probably extend it (all byte-compiled functions are functions).

Really what I have as `elsa-function-type` should maybe be called "callable" or "invokable" or something like that.

Reply:
I removed it from my commit. We can improve/rename the function type later.